### PR TITLE
ref #236: correct total value on percentage discounts

### PR DESCRIPTION
**Pull request by pSpeck**
| Q             | A
| ------------- | ---
| Branch        | 6.2.x
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes 
| Fixed issues  | chameleon-system/chameleon-system#236
| License       | MIT

### CODE REVIEW

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 1]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
+        $this->correctBasketItemDiscountPrices($articlesAffected, $dVoucherValue, $totalValueOfAffectedItems);
+    }
 
-        $iNumberOfAffectedItems = count($aArticlesAffected);
-        $dTotalValueOfAffectedItemsUsed = 0;
+    /**
+     * @param TdbShopDiscount $discount
+     * @param float           $unusedDiscountValue value difference between discount based on single article and discount based on all articles
+     */
+    public function ReducePriceForItemsAffectedByDiscount(\TdbShopDiscount $discount, float $unusedDiscountValue): void
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
    public function reducePriceForItemsAffectedByDiscount(\TdbShopDiscount $discount, float $unusedDiscountValue): void
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 2]
```diff
@@ -617,12 +639,19 @@ public function ApplyDiscount(TdbShopDiscount $oDiscount)
...
                 $oDiscountToPass = clone $oDiscount;
-                $dTotalDiscountValue = $oItem->ApplyDiscount($oDiscountToPass, $dTotalDiscountValue);
+                $oItem->ApplyDiscount($oDiscountToPass, $dTotalDiscountValue);
+                $totalDiscountValueItemCalculated += $oDiscountToPass->dRealValueUsed;
             }
         }
+        $totalDiscountValueOverAll = $oDiscount->GetValue();
+        if ($totalDiscountValueOverAll !== $totalDiscountValueItemCalculated) {
+            $missingDiscountValue = $totalDiscountValueOverAll - $totalDiscountValueItemCalculated;
+            $this->ReducePriceForItemsAffectedByDiscount($oDiscount, $missingDiscountValue);
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
            $this->reducePriceForItemsAffectedByDiscount($oDiscount, $missingDiscountValue);
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 3]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
+     * @param TdbShopDiscount $discount
+     * @param float           $unusedDiscountValue value difference between discount based on single article and discount based on all articles
+     */
+    public function ReducePriceForItemsAffectedByDiscount(\TdbShopDiscount $discount, float $unusedDiscountValue): void
+    {
+        $currentPosition = $this->getItemPointer();
+        $this->GoToStart();
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
        while ($basketItem = $this->Next()) {
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 4]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
-        $aArticlesAffected = array();
-        $dTotalValueOfAffectedItems = 0;
-        while ($oItem = &$this->Next()) {
-            $bIncludeProduct = $oVoucher->AllowVoucherForArticle($oItem);
-            if ($bIncludeProduct) {
-                $aArticlesAffected[] = &$oItem;
-                $dTotalValueOfAffectedItems = $dTotalValueOfAffectedItems + $oItem->dPriceTotalAfterDiscountWithoutVouchers;
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
        while ($basketItem = $this->Next()) {
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 5]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
+     */
+    public function ReducePriceForItemsAffectedByDiscount(\TdbShopDiscount $discount, float $unusedDiscountValue): void
+    {
+        $currentPosition = $this->getItemPointer();
+        $this->GoToStart();
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
+            if (true === $discount->AllowDiscountForArticle($basketItem)) {
+                $articlesAffected[] = &$basketItem;
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
                $articlesAffected[] = $basketItem;
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 6]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
-        while ($oItem = &$this->Next()) {
-            $bIncludeProduct = $oVoucher->AllowVoucherForArticle($oItem);
-            if ($bIncludeProduct) {
-                $aArticlesAffected[] = &$oItem;
-                $dTotalValueOfAffectedItems = $dTotalValueOfAffectedItems + $oItem->dPriceTotalAfterDiscountWithoutVouchers;
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
+            if (true === $oVoucher->AllowVoucherForArticle($basketItem)) {
+                $articlesAffected[] = &$basketItem;
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]:
```suggestion
                $articlesAffected[] = $basketItem;
```

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 7]
```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
...
+    }
+
+    protected function correctBasketItemDiscountPrices(array $articlesAffected, float $unusedDiscountValue, float $totalValueOfAffectedItems): void
+    {
+        $iNumberOfAffectedItems = count($articlesAffected);
         if ($iNumberOfAffectedItems > 0) {
-            $dRemainingVoucherUse = $dVoucherValue;
-            foreach (array_keys($aArticlesAffected) as $iAffectedArticleIndex) {
+            $dRemainingDiscountUse = $unusedDiscountValue;
+            foreach (array_keys($articlesAffected) as $iAffectedArticleIndex) {
```
Comment (umulmrum) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: This method is really hard to read. Please improve it by
a) adding an early return if iNumberOfAffectedItems <= 0
b) replacing `foreach (array_keys($articlesAffected) as $iAffectedArticleIndex)` by `foreach ($articlesAffected as $iAffectedArticleIndex => $affectedArticle)` (and of course use $affectedArticle)
c) splitting the method in at least 3 methods

Thanks!

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 1 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: changed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 2 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: changed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 3 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: removed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 4 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: removed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 5 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: removed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 6 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: removed

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 7 above)
Comment (pSpeck) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: imporve method realising a and b

**Comment by umulmrum** [comment]
Thanks @pSpeck

**Pull request merged**